pkg/ingester/index.go

```diff
@@ -40,56 +51,73 @@ func (i *invertedIndex) add(metric model.Metric, fp model.Fingerprint) {
 	}
 }
 
-func (i *invertedIndex) lookup(matchers []*labels.Matcher) []model.Fingerprint {
+func (ii *invertedIndex) lookup(matchers []*labels.Matcher) []model.Fingerprint {
 	if len(matchers) == 0 {
 		return nil
 	}
-	i.mtx.RLock()
-	defer i.mtx.RUnlock()
 
-	// intersection is initially nil, which is a special case.
-	var intersection []model.Fingerprint
-	for _, matcher := range matchers {
-		values, ok := i.idx[model.LabelName(matcher.Name)]
-		if !ok {
-			return nil
-		}
-		var toIntersect []model.Fingerprint
-		if matcher.Type == labels.MatchEqual {
-			fps := values[model.LabelValue(matcher.Value)]
-			toIntersect = merge(toIntersect, fps)
-		} else {
-			for value, fps := range values {
-				if matcher.Matches(string(value)) {
-					toIntersect = merge(toIntersect, fps)
+	result := []model.Fingerprint{}
+
+outer:
+	for i := range ii.shards {
+		ii.shards[i].mtx.RLock()
```

Comment:
Is it worth assigning `ii.shards[i]` to something?

Reply:
Actually broke them out into separate functions, as the locking was wrong.